feat(frontend): improve accessibility of dashboard components

zwliew opened the pull request:
## Problem

We want to implement frontend tests soon. However, some dashboard elements are inaccessible to the screen readers. We need to make these elements accessible so that the test environment can easily query them for testing.

## Solution

Where possible, replace generic elements like `div` with semantic elements like `button`. Otherwise, attach appropriate ARIA roles to the elements.

These changes allow the test environment to easily query significant elements by role and/or label.
 
**Improvements**:

- Add the `title` prop to `CloseButton`
- Convert `CloseButton` to an actual `button` (and styled it properly)
- Convert credential input buttons to `TextButton` components (and styled them properly)
- Add labels to the input boxes in the campaign creation pages
- Add semantically appropriate ARIA roles to some custom components:
  - `Checkbox`: role="checkbox"
  - `Dropdown`: role="listbox" (and role="option" for dropdown options)
  - `FileInput` button: role="button"
  - `SendRate` header button: role="button"
  - `ErrorBlock`: role="alert"

ARIA role reference: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles

## Before & After Screenshots

**1. A tooltip appears when user hovers over close button**
Before:
![Screenshot 2021-03-26 at 6 45 02 PM](https://user-images.githubusercontent.com/4538946/112620361-63ab2880-8e63-11eb-9e1e-ade75cb61af7.png)

After:
![Screenshot 2021-03-26 at 6 44 23 PM](https://user-images.githubusercontent.com/4538946/112620329-57bf6680-8e63-11eb-94e7-68176802e8c9.png)


**2. The credential input buttons are 16px in size (matching the Figma design)**

Before:
![Screenshot 2021-03-26 at 6 45 30 PM](https://user-images.githubusercontent.com/4538946/112620413-77568f00-8e63-11eb-856c-9c49f3c9cc85.png)

After:
![Screenshot 2021-03-26 at 6 47 12 PM](https://user-images.githubusercontent.com/4538946/112620580-b4bb1c80-8e63-11eb-8577-52dcda6c389f.png)

## Tests

These tests have been verified locally, but not on the Amplify staging endpoint yet.

1. Compile and deploy the app on staging.
2. Conduct smoke tests.
3. The app should look and function exactly like before, but the elements should have additional role attributes. Also, some labels are now clickable and focus on the relevant textboxes when clicked.

Smoke tests:
 - Creating and sending email campaign
 - Creating and sending SMS campaign
 - Creating and sending Telegram campaign
 - Creating and sending protected email campaign
 - Viewing and decrypting protected email page
 - Unsubscribing

## Deploy Notes

There are no deployment changes.

zwliew commented on frontend/src/components/common/file-input/FileInput.tsx:
```diff
@@ -24,7 +24,7 @@ const FileInput = ({
         onChange={(e) => onFileSelected(e.target.files)}
       />
-      <label htmlFor="recipient-upload-input">
+      <label role="button" htmlFor="recipient-upload-input">
```
I considered that at first, but I realised that the button that the user actually sees and clicks on is the `label`, so I added the role to the label instead.

miazima replied:
Ah I see what you mean 👍

miazima commented:
![image](https://user-images.githubusercontent.com/5744384/112972221-9e260580-9182-11eb-873b-60dc7297af83.png)

lgtm, just one issue where there is a border introduced on `:focus`, when mouse down on the create campaign close button

zwliew commented:
Good catch. I'm not sure if we should remove it though.

It only appears when users use non-mouse devices to focus on it, like a keyboard. In such cases, the outline provides visual confirmation that the button is being focused on. I think it is good for accessibility.

When focusing on the button with a mouse, the outline doesn't appear, so it doesn't affect that use case.

miazima commented:
> When focusing on the button with a mouse, the outline doesn't appear, so it doesn't affect that use case.

When I click the close button using the trackpad it does appear though 🤔 
I understand that it helps on the accessibility part, is there a way we can disable the focus for normal trackpad/mouse usage?

zwliew commented:
> > When focusing on the button with a mouse, the outline doesn't appear, so it doesn't affect that use case.
> 
> When I click the close button using the trackpad it does appear though 🤔
> I understand that it helps on the accessibility part, is there a way we can disable the focus for normal trackpad/mouse usage?

Oh I see! Somehow I couldn't reporoduce it on Chrome Canary, but I could on Chrome. I'll look into it!

zwliew commented:
I've rebased the branch and added 2 commits to fix some issues, including the issue you mentioned. Now the outline only appears when using a keyboard.

miazima commented:
lgtm!

zwliew commented:
Thanks for the review!
